Добавляет вариант дял препроцессора Sass(Scss)

dimaos32 opened the pull request:
В файле nebo.scss в классе .nebo сохранены исходные css-переменные по аналогии с файлом nebo.css Можно использовать все, что доступно в классе nebo.css - переопределять css-переменные, использовать модификаторы Так же можно использовать миксин apply-nebo-corner() - без аргументов он принимает значения по-умолчанию (как в файле nebo.css) Миксин принимает 4 аргумента - $r (радиус), $w (ширина), $h (высота) и $corner (угол - строка, возможные варианты: "tl", "tr", "bl" и "br")

dimaos32 committed:
Добавляет вариант дял препроцессора Sass(Scss)

В файле nebo.scss в классе .nebo сохранены исходные css-переменные по аналогии с файлом nebo.css
Можно использовать все, что доступно в классе nebo.css - переопределять css-переменные, использовать модификаторы
Так же можно использовать миксин apply-nebo-corner() - без аргументов он принимает значения по-умолчанию (как в файле nebo.css)
Миксин принимает 4 аргумента - $r (радиус), $w (ширина), $h (высота) и $corner (угол - строка, возможные варианты: "tl", "tr", "bl" и "br")

dimaos32 commented:
Все аргументы в миксине необязательные - для неуказанных берутся значения по-умолчанию

polyuretanio commented:
@dimaos32 только добрался. я правильно понимаю, что реализация миксина не позволит на лету переопределять "внутренние переменные" библиотеки. чтобы например реализовать кейс с адаптивностью? Вот как на этом шаге демки —https://htmlacademy.ru/demos/197#67

И можно ли доработать миксин таким образом, чтобы все фичи нативной реализации сохранялись и в миксине?

dimaos32 commented:
Нет - очень даже позволит.
Миксин специально писался так, чтобы если подключать scss-версию - код, написанный для css-версии работал аналогично.
То есть у пользователя есть два варианта, как использовать библиотеку:
- переназначать кастомные переменные (нативный css способ)
- использовать миксин (возможность препроцессора sass(scss)
Вот [демо](https://codepen.io/dimaos/pen/wBBjmVo) - пример использования - 160 и 166 строка

dimaos32 commented:
@polyuretanio Нет - очень даже позволит.
Миксин специально писался так, чтобы если подключать scss-версию - код, написанный для css-версии работал аналогично.

То есть у пользователя есть два варианта, как использовать библиотеку:
- переназначать кастомные переменные (нативный css способ)
- использовать миксин (возможность препроцессора sass(scss)

Вот [демо](https://codepen.io/dimaos/pen/wBBjmVo) - пример использования - 160 и 166 строка

dimaos32 commented:
Так же первой карточке (пример с кастомными переменными) добавлен модификатор угла в верстке.

dimaos32 commented:
Единственное ограничение - если использовать миксин - класс-модификатор (для угла) игнорируется, даже если в миксин не передать переменную `$corner` - применится значение по-умолчанию (правый нижний угол).
Отсюда вытекает рекомендация - использовать либо css-стиль(переопределение кастомные переменных и для угла класс-модификатор), либо миксин. Технически можно переопределять кастомные переменные, а в миксин передавать только угол - но мне такое решение кажется слишком запутанным.

dimaos32 commented:
@polyuretanio - ты прав, я перечитал вопрос и пересмотрел демку - да, в моем решении все промежуточные переменные удалены, соответственно адаптив (разные углы) возможен только через миксин.

Концептуально это не правильно - надо оставить весь оригинальный функционал.
Я на днях посмотрю, что можно сделать и добавлю коммит.

dimaos32 committed:
Добавляет совместимость scss версии с css версией

Добавляет поддержку всех, в том числе и промежуточных кастомных переменных из css версии.

[Обновленное демо](https://codepen.io/dimaos/pen/wBBjmVo)

dimaos32 commented:
@polyuretanio добавил совместимость. Все кастомные переменные из css версии сохранены.

[Обновленное демо](https://codepen.io/dimaos/pen/wBBjmVo)

polyuretanio commented:
@dimaos32 отлично. я правильно понимаю, что препроцессорный код, в основном, начинается с 85 строчки? я просто обновил саму библиотеку, добавив в неё 6 доп. параметров для тонкой настройки углов. то есть поменялся код внутри основного css-правила .nebo. по идее, нетрудно будет твоё решение обновить?

dimaos32 commented:
@polyuretanio верно - препроцессорный код это только миксин, который переопределяет кастомные переменные.
Я сегодня-завтра могу посмотреть твои обновления и дополнить препроцессорный вариант.

dimaos32 committed:
Add custom corners support and different radius values for x & y axes

dimaos32 commented:
@polyuretanio добавил поддержку разных углов и разных значений для радиуса по осям.
В миксин можно передать как и общий радиус, так и конкретные радиусы (для конкретного угла по конкретной оси). Сначала устанавливается значение общего радиуса (если соответствующий аргумент передан в миксин), а потом те, что заданы точечно - перезаписывают соответствующие кастомки.

dimaos32 commented:
Положительное отличие от css-версии - если из радиусов в миксин передать только общий радиус - то частные радиусы перезапишутся (если переопределить только кастомку общего радиуса - точечные не обновятся - и я не знаю, как это решить на чистом css).

dimaos32 commented:
Беря во внимание предыдущее сообщение - я бы добавил промежуточную переменную с радиусом (одинаковым по обеим осям) для каждого угла - т.к. вероятность, что понадобятся разные радиусы для каждого закругления высокая, а от, что скругление будет эллипсом - сильно ниже.
Примерно так:
```css
  --nb-r: 20px;
  --nb-w: 20px;
  --nb-h: 20px;

  /* Fine-tuned radiuses */
  --nb-cor1-r: var(--nb-r);
  --nb-cor2-r: var(--nb-r);
  --nb-curve-r: var(--nb-r);

  --nb-cor1-rw: var(--nb-cor1-r);
  --nb-cor1-rh: var(--nb-cor1-r);
  --nb-cor2-rw: var(--nb-cor2-r);
  --nb-cor2-rh: var(--nb-cor2-r);
  --nb-curve-rw: var(--nb-curve-r);
  --nb-curve-rh: var(--nb-curve-r);
```
П.С. Мне почему-то кажется, что постфиксы `rw` и `rh` можно заменить на `rx` и `ry` соответственно - как буд-то так больше ясности.

П.П.С. Могу сам это написать для обеих версий (ванильной и препроцессорной) - но наверное лучше когда добьем и примем текущий вариант препроцессорной версии.

polyuretanio commented:
@dimaos32 извини, надолго выпал, уезжал в отпуск

> Положительное отличие от css-версии - если из радиусов в миксин передать только общий радиус - то частные радиусы перезапишутся (если переопределить только кастомку общего радиуса - точечные не обновятся - и я не знаю, как это решить на чистом css).

В CSS-версии это решено достаточно просто — все кастомные углы берут значение общего радиуса, и если какой-то переопределяется явно, то остальные остаются как были. То есть общий радиус устанавливается как общий дефолт и точечно переопределяется там, где надо

> П.С. Мне почему-то кажется, что постфиксы rw и rh можно заменить на rx и ry соответственно - как буд-то так больше ясности.

Я решил сохранить консистентность с параметрами `--nb-w` и `--nb-h`, где w отсылает к ширине (горизонтали), а h к высоте (вертикали)

Я тогда мёржу препроцессорный вариант
